polyfill/index.d.ts

```diff
@@ -0,0 +1,385 @@
+     *
+     * The default depends on the type being used.
+     *
+     * TODO: document which types use which defaults.
+     */
+    largestUnit: T;
+  }
+
+  export type DurationLike = {
+    years?: number;
+    months?: number;
+    days?: number;
+    hours?: number;
+    minutes?: number;
+    seconds?: number;
+    milliseconds?: number;
+    microseconds?: number;
+    nanoseconds?: number;
+  };
+
+  export class Duration extends Required<DurationLike> {
+    static from(
+      item: Temporal.Duration | DurationLike | string | object,
+      options?: AssignmentOptions
+    ): Temporal.Duration;
+    constructor(
+      years?: number,
+      months?: number,
+      days?: number,
+      hours?: number,
+      minutes?: number,
+      seconds?: number,
+      milliseconds?: number,
+      microseconds?: number,
+      nanoseconds?: number
+    );
+    readonly years: number;
+    readonly months: number;
+    readonly days: number;
+    readonly hours: number;
+    readonly minutes: number;
+    readonly seconds: number;
+    readonly milliseconds: number;
+    readonly microseconds: number;
+    readonly nanoseconds: number;
+    with(durationLike: DurationLike, options: AssignmentOptions): Temporal.Duration;
+    plus(other: Temporal.Duration, options: ArithmeticOptions): Temporal.Duration;
+    minus(other: Temporal.Duration, options: DurationMinusOptions): Temporal.Duration;
+    getFields(): Required<DurationLike>;
+    toLocaleString(locales?: string | string[], options?: Intl.DateTimeFormatOptions): string;
+    toJSON(): string;
+    toString(): string;
+  }
+
+  export class Absolute extends Required<AbsoluteLike> {
+    static fromEpochSeconds(epochSeconds: number): Temporal.Absolute;
+    static fromEpochMilliseconds(epochMilliseconds: number): Temporal.Absolute;
+    static fromEpochMicroseconds(epochMicroseconds: bigint): Temporal.Absolute;
+    static fromEpochNanoseconds(epochNanoseconds: bigint): Temporal.Absolute;
+    static from(item: Temporal.Absolute | string | object): Temporal.Absolute;
+    static compare(one: Temporal.Absolute, two: Temporal.Absolute): 1 | -1 | 0;
+    constructor(epochNanoseconds: bigint);
+    getEpochSeconds(): number;
+    getEpochMilliseconds(): number;
+    getEpochMicroseconds(): bigint;
+    getEpochNanoseconds(): bigint;
+    plus(temporalDurationLike: DurationLike): Temporal.Absolute;
+    minus(temporalDurationLike: DurationLike): Temporal.Absolute;
+    difference(
+      other: Temporal.Absolute,
+      options?: DifferenceOptions<'days' | 'hours' | 'minutes' | 'seconds'>
+    ): Temporal.Duration;
+    // Should Absolute and DateTime have different names for `inTimeZone`? See #574.
+    inTimeZone(temporalTimeZoneLike?: TimeZoneLike): Temporal.DateTime;
+    toLocaleString(locales?: string | string[], options?: Intl.DateTimeFormatOptions): string;
+    toJSON(): string;
+    toString(temporalTimeZoneLike?: TimeZoneLike): string;
+  }
+
+  export type DateLike = {
```

Comment:
Not necessarily within scope for this pull request, but for calendar-aware types (Date, DateTime, YearMonth, MonthDay) this is about to get a bit more complicated. A few calendars are going to require extra fields (such as `era` for the Japanese imperial calendar) and those will be passed to `from() / with()` and returned from `getFields()`.

Reply:
Haha, I asked the same question in my comment above! I'll delete it there and move that discussion here. 

> Hey, will calendars expose custom fields, like Japanese `era` on `Date`?  If yes, then the way we'd probably solve this in TS is to make types generic, e.g. `DateTimeLike<CalendarType = Temporal.IsoCalendar>` or something like that. This would enables calendars to expose extra fields to the calendar types while still retaining type safety, and would let users assert that they're expecting dates from a particular calendar to have particular fields.

Anyway, where will the extra fields be? On which types?  And what about the Calendar type itself-- where is it documented?

Reply:
Oops, I must have missed that comment! Sorry!

The extra fields can potentially be on Date, DateTime, YearMonth, and MonthDay. The Calendar type (draft API) is currently documented in https://tc39.es/proposal-temporal/docs/calendar-draft.html and there's also ongoing implementation work in progress at https://github.com/tc39/proposal-temporal/commits/calendars

Reply:
OK. Do you think I should add Calendar to TS types now, or wait for the Calendar API to land and make changes then?

Reply:
I think it would be better to wait for the Calendar API to land. (Basically as soon as the TS types land, we should require that any pull requests that make subsequent API changes also update the types file.)